llmfoundry/callbacks/hf_checkpointer.py

```diff
@@ -220,9 +221,10 @@ def _save_checkpoint(self, state: State, logger: Logger):
                 # TODO: after torch 2.1, we can load a state dict into a meta model
                 # and skip the extra model init
                 log.debug(f'Creating new model instance')
-                new_model_instance = type(original_model)(copied_config)
-                new_model_instance.to(dtype=self.dtype)
-                new_model_instance.load_state_dict(state_dict)
+                with init_empty_weights():
+                    new_model_instance = type(original_model)(copied_config)
+
+                new_model_instance.load_state_dict(state_dict, assign=True)
```

Comment:
would we have been able to previously do:
```
                with init_empty_weights():
                    new_model_instance = type(original_model)(copied_config)

                new_model_instance.load_state_dict(state_dict)

```

Reply:
`assign=False` means that the tensors in the model retain the properties of the tensors in the model, rather than the tensors in the state dict. Since the tensors in the model are on meta device, they would stay on meta device, and the load state dict would not succeed.

Reply:
e.g.
```
In [1]: from llmfoundry.models.utils import init_empty_weights

In [2]: import torch

In [3]: l1 = torch.nn.Linear(4,4)

In [4]: sd = l1.state_dict()

In [5]: with init_empty_weights():
   ...:     l2 = torch.nn.Linear(4,4)
   ...:

In [6]: l2.load_state_dict(sd)
/Users/daniel.king/miniconda3/envs/foundry-3.10/lib/python3.10/site-packages/torch/nn/modules/module.py:2025: UserWarning: for weight: copying from a non-meta parameter in the checkpoint to a meta parameter in the current model, which is a no-op. (Did you mean to pass `assign=True` to assign items in the state dictionary to their corresponding key in the module instead of copying them in place?)
  warnings.warn(f'for {key}: copying from a non-meta parameter in the checkpoint to a meta '
/Users/daniel.king/miniconda3/envs/foundry-3.10/lib/python3.10/site-packages/torch/nn/modules/module.py:2025: UserWarning: for bias: copying from a non-meta parameter in the checkpoint to a meta parameter in the current model, which is a no-op. (Did you mean to pass `assign=True` to assign items in the state dictionary to their corresponding key in the module instead of copying them in place?)
  warnings.warn(f'for {key}: copying from a non-meta parameter in the checkpoint to a meta '
Out[6]: <All keys matched successfully>

In [7]: l2.weight.device
Out[7]: device(type='meta')
```

Reply:
Can we put a comment here describing that it is initing on meta and loading as you described? wasn't clear to me either until your description

Reply:
added